### DISCUSSION TITLE
Logging into Git as a GitHub App

### DISCUSSION THREAD

**Discussion by cjdenio** [community/community#24664] [Apps, API and Webhooks]
<p>How would I log into Git as a GitHub App? :thinking:  I’ve sort of got it working, but I’m curious what the “actual” way of doing this is :smiley:  What would I set as the Git username and email? Thank you!</p>

**Comment by francisfuzz** [comment, marked as answer]
<p>:wave: <a class="mention" href="/cjdenio">@cjdenio</a> –– great questions! Thank <em>you</em> for raising them here :bowing_man:</p>
<blockquote>
<p><em>How would I log into Git as a GitHub App? :thinking: I’ve sort of got it working, but I’m curious what the “actual” way of doing this is :smiley:</em></p>
</blockquote>
<p>It’s possible to <a href="https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/">authenticate with GitHub.com as a GitHub App</a>; here’s the <a href="https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation">section on HTTP-based Git access by an installation</a>.</p>
<blockquote>
<p><em>What would I set as the Git username and email? Thank you!</em></p>
</blockquote>
<p>That’s a great question! The way we currently attribute commits made by GitHub Apps isn’t currently publicly documented in our <a href="https://help.github.com/en/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address"><code>Setting your commit email address</code></a> article and is subject to change at <strong><em>anytime without notice</em></strong>. If it’s helpful in the meantime, here’s something you could try (though your mileage may vary over time :v:).</p>
<p>You could set the Git username is the name of your application with <code>[bot]</code> appended at the end. For example, if your bot’s name is <code>machina</code>, then the Git username is <code>machina[bot]</code>:</p>
<pre><code class="lang-auto">git config user.name "machina[bot]"
</code></pre>
<p>For the email, you could set it to the GitHub App’s user <code>id</code>, a plus sign (<code>+</code>), and its name with <code>[bot]</code> appended, followed by the Users No Reply GitHub email address. For example, if the app’s user <code>id</code> is <code>123456789</code>, then the email would be:</p>
<pre><code class="lang-auto">git config user.email "123456789+machina[bot]@users.noreply.github.com"
</code></pre>
<p>Here’s an example commit showing the GitHub Actions installation committing to a repository: <a href="https://github.com/actions-packages-examples/github-script__playground/commit/f75a13b585d89017ab76a163e3fade61a051420f"><code>actions-packages-examples/github-script__playground@f75a13b</code></a></p>
<p>In its <a href="https://github.com/actions-packages-examples/github-script__playground/commit/f75a13b585d89017ab76a163e3fade61a051420f.patch">patch file</a>, observe the second line’s contents:</p>
<pre><code class="lang-auto">From: "github-actions[bot]"
 &lt;41898282+github-actions[bot]@users.noreply.github.com&gt;
</code></pre>
<p>One approach for obtaining the GitHub App’s user ID is to <a href="https://developer.github.com/v3/repos/contents/#create-or-update-file-contents">create a new file using the Contents API</a>, navigating to the commit on <a href="http://GitHub.com">GitHub.com</a>, and appending <code>.patch</code> to the end of that commit URL to reveal the patch information showcasing the actor name and email address we use. I hope this helps!</p>

**Reply by cmorinupgrade** [reply]
> For the email, you could set it to the GitHub App’s user id, a plus sign (+), and its name with [bot] appended, followed by the Users No Reply GitHub email address. For example, if the app’s user id is 123456789, then the email would be:

Not to be confused with GitHub App ID. This must be the UID for the GitHub App itself. You can find it using the approach from @alcore : https://github.com/orgs/community/discussions/24664#discussioncomment-3880274

**Reply by AlissonRS** [reply]
@francisfuzz the commit you shared results in a 404. In short, I was looking for how to require pull requests before merging into `main` branch, but allow a commit to be done via GitHub Action without a PR (for example by using GitHub App and allowing it to bypass the PR requirement).

**Reply by lindhe** [reply]
Does anyone know what the noreply email looks like for GitHub Enterprise? This seems to not be documented at all…

![image](https://github.com/community/community/assets/7773090/71fc17b9-cf99-4e99-9167-b78e6985e391)

**Comment by cjdenio** [comment]
<p>Wow, this is is awesome! :exploding_head: Thanks for taking the time to answer; I really appreciate having closure on this question :smiley:</p>

**Comment by inez** [comment]
<p><a class="mention" href="/francisfuzz">@francisfuzz</a> Is there a way to authenticate with Github over SSH as a Github App - maybe somehow using Github App private key?<br>
And by over SSH I meant to be able to push and pull code, not any shell access (obviously).</p>

**Comment by francisfuzz** [comment]
<blockquote>
<p>Is there a way to authenticate with Github over SSH as a Github App - maybe somehow using Github App private key?<br>
And by over SSH I meant to be able to push and pull code, not any shell access (obviously).</p>
</blockquote>
<p>:wave: <a class="mention" href="/inez">@inez</a>: Great question! There’s no way for a GitHub App to authenticate using an SSH key at this time.</p>

**Comment by msgongora** [comment]
<p>I was so happy to read this entry, thank you for sharing it <a class="mention" href="/francisfuzz">@francisfuzz</a>. I’d like to mention that you could get the user id from the Content API response, same endpoint you referenced: <a href="https://docs.github.com/en/rest/reference/repos#create-or-update-file-contents" rel="noopener nofollow ugc">create new file</a>.</p>

**Comment by hsy3418** [comment]
<p><a class="mention" href="/francisfuzz">@francisfuzz</a> thanks a lot for this information! I have one more question, how did you make the bot has the verified signature in this <a href="https://github.com/actions-packages-examples/github-script__playground/commit/f75a13b585d89017ab76a163e3fade61a051420f" class="inline-onebox" rel="noopener nofollow ugc">updated: 1591652451925 · actions-packages-examples/github-script__playground@f75a13b · GitHub</a>, I am trying to do similar things, authenticate the github app in the github actions, and make a new commit, but I couldn’t make the bot’s signature to be verified.</p>

**Comment by francisfuzz** [comment]
<p><a class="mention" href="/hsy3418">@hsy3418</a> - For that particular instance, I left a comment here:</p>
<aside class="onebox githubcommit">
  <header class="source">
      <a href="https://github.com/actions-packages-examples/github-script__playground/commit/f75a13b585d89017ab76a163e3fade61a051420f#commitcomment-50944805" target="_blank" rel="noopener">github.com/actions-packages-examples/github-script__playground</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Commit">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M10.86 7c-.45-1.72-2-3-3.86-3-1.86 0-3.41 1.28-3.86 3H0v2h3.14c.45 1.72 2 3 3.86 3 1.86 0 3.41-1.28 3.86-3H14V7h-3.14zM7 10.2c-1.22 0-2.2-.98-2.2-2.2 0-1.22.98-2.2 2.2-2.2 1.22 0 2.2.98 2.2 2.2 0 1.22-.98 2.2-2.2 2.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/actions-packages-examples/github-script__playground/commit/f75a13b585d89017ab76a163e3fade61a051420f" target="_blank" rel="noopener">updated: 1591652451925</a>
    </h4>

    <div class="github-info">
      <div class="date">
        committed <span class="discourse-local-date" data-format="ll" data-date="2020-06-08" data-time="21:40:52" data-timezone="UTC">09:40PM - 08 Jun 20 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/apps/github-actions" target="_blank" rel="noopener">
          <img alt="github-actions[bot]" src="https://user-images.githubusercontent.com/15894826/181085935-b738d7b6-b3c8-4780-bc2d-f04bd0910ca3.png" class="onebox-avatar-inline" width="20" height="20">
          github-actions[bot]
        </a>

      </div>

      <div class="lines" title="changed 1 files with 1 additions and 0 deletions">
        <a href="https://github.com/actions-packages-examples/github-script__playground/commit/f75a13b585d89017ab76a163e3fade61a051420f" target="_blank" rel="noopener">
          <span class="added">+1</span>
          <span class="removed">-0</span>
        </a>
      </div>
    </div>

  </div>
</div>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by twebner** [comment]
<p>Hi <a class="mention" href="/francisfuzz">@francisfuzz</a>, the link for bot signature verification doesn’t work anymore. Can you please share a new one or explain how to get verified signatures when using a GitHub app token to push commits in a workflow?</p>

**Comment by alcore** [comment]
A much simpler approach to getting your app's "User ID" is to simply call:

https://api.github.com/users/myapp[bot]

Replacing `myapp` with your app name, but leaving the `[bot]` suffix in place. You can get any app's user ID this way (yes, even private apps) - you do not even need to use an auth token for that request, let alone wrangle with creating a JWT from your private key and using the contents API to make a commit from which to get that ID.

**Reply by lindhe** [reply]
I'm guessing it's https://github.example.com/api/v3/users/myapp[bot] for enterprise servers?

**Reply by joshjohanning** [reply]
@lindhe yes! 

If you call via `curl` though, you have to encode the URL, like this: 

```
curl -L  -H "Authorization: Bearer <YOUR-TOKEN>" https://github.mycompany.com/api/v3/users/my-app%5Bbot%5D
```

And then for the email address, you have to use your domain in the email address, such as 

```
123456+my-app[bot]@users.noreply.github.mycompany.com
```

**Comment by mkushakov** [comment]
Hello,

Thank you for explanation. We was able to push commit with GitHub App from our Jenkins. 
But one question still remain:
- Did you able to use GitHub App so push back verified (signed) commits?
All links are quite old and point to 404 :(

**Reply by HenryNguyen5** [reply]
There's a good document here on how to create signed commits using a Github App.
- https://github.com/Nautilus-Cyberneering/pygithub/blob/main/docs/how_to_sign_automatic_commits_in_github_actions.md

## TL;DR:
- You can use either the GraphQL API to more(ish) easily create signed commits https://docs.github.com/en/graphql/reference/mutations#commit-signing
- Or you can use Github's REST api to create signed commits, if you want to create a single commit file, [it's pretty easy](https://gist.github.com/swinton/03e84635b45c78353b1f71e41007fc7c), but it gets pretty involved if you want to [create a multi-file commit](https://github.com/Nautilus-Cyberneering/pygithub/blob/main/docs/how_to_create_a_single_commit_with_multiple_files_using_github_api.md#solution-01-using-gitHub-rest-api)

**Reply by mkushakov** [reply]
We end up creating our own GO CLI to be able to create signed commits in Jenkins - https://github.com/nexthink-oss/ghup

**Reply by HenryNguyen5** [reply]
That looks great! When your CLI creates tags / annotated tags via the github UI, are they automatically signed too?

**Reply by isometry** [reply]
Unfortunately, the GitHub API does not sign annotated tags. Lightweight tags ("releases" in GitHub's terminology) are just references to the target commit, so inherit the target commit's signature.

**Reply by HenryNguyen5** [reply]
Ah yes, I ended up testing this myself too. Quite odd considering that an annotated tag is almost like a commit, which works with GH's API for generating signed commits.
